testkit/src/api.rs

```diff
@@ -311,6 +307,96 @@ impl TestKitApi {
     }
 }
 
+#[derive(Debug)]
+enum ApiAccess {
+    Public,
+    Private,
+}
+
+impl fmt::Display for ApiAccess {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            ApiAccess::Public => formatter.write_str("public"),
+            ApiAccess::Private => formatter.write_str("private"),
+        }
+    }
+}
+
+// Logging middleware for `TestKitApi`.
+fn log_request(
+    access: &ApiAccess,
+    request: &mut Request,
+    mut response: Response,
+) -> IronResult<Response> {
+    use iron::headers::ContentLength;
+    use iron::method::Method;
+    use iron::response::WriteBody;
+
+    fn has_body(method: &Method) -> bool {
+        match *method {
+            Method::Get | Method::Head | Method::Delete | Method::Trace => false,
+            _ => true,
+        }
+    }
+
+    if !log_enabled!(Level::Trace) {
+        // Avoid expensive string allocations.
+        return Ok(response);
+    }
+
+    let mut url = request.url.path().join("/");
+    if let Some(query) = request.url.query() {
+        url += "?";
+        url += query;
+    }
+
+    let req_body: Option<String> = if has_body(&request.method) {
+        request
+            .get::<bodyparser::Raw>()
+            .expect("cannot read request body")
+    } else {
+        None
+    };
+
+    let response_body: Option<String> = {
+        let len: u64 = response
+            .headers
+            .get::<ContentLength>()
+            .map(|len| len.0)
+            .unwrap_or_default();
+
+        response.body.take().map(|mut body| {
+            let mut buffer = Vec::with_capacity(len as usize);
+            body.write_body(&mut buffer)
+                .expect("cannot write response body to buffer");
+            String::from_utf8(buffer).expect("cannot decode response body")
+        })
+    };
+
+    trace!(
+        "{method} ({access}) /{url}{req_body}\nResponse: {resp_status}\n{resp_body}\n",
+        method = request.method,
+        access = access,
+        url = url,
+        req_body = req_body
+            .map(|body| format!("\nBody: {}", body))
```

Comment:
nitpick: I'm not a fan of avoiding **all** allocations in code, but you can easily avoid any of them in this part of the code (by the price of less readable message in case of no request body).